### ISSUE TITLE
Support hstore in postgis plugin

### ISSUE THREAD

**Issue by herm** [mapnik/mapnik#1113]
It would be nice if postgis was able to read an hstore value.
Use case: pass this value to a metawriter and write all values defined in this hstore.

**Comment by TehNrd** [comment]
Agreed this should support more than just hstore, especially with the more robust JSON support in PostgreSQL 9.4 .

Here is a use case. The census provides a ton of data points (1000s) for a given geometry. Rather than pollute my tables with thousands of columns (would hit Postgres limits anyway) I was hoping to create on JSON column (censusvars) and populate with data like so:

`
{"DP02_0068PE":18.5,  
"DP04_0009E":4232,  
"DP02_0069E":4353454,  
...
}
`

The style would look like this:

`<Filter>[censusvars->'DP02_0001E'] &gt; 0</Filter>`

**Comment by TehNrd** [comment]
I wonder if it worth having a separate issue created for JSON/B support. Thoughts?

**Comment by springmeyer** [comment]
@TehNrd could you create a small testcase demonstrating what you'd like to do? Also: one reason this has not moved is that a good workaround is to use a SQL subselect with mapnik that transforms any custom fields into a normal column on the fly.

**Comment by TehNrd** [comment]
@springmeyer Ah, thank you. Sub-query is exactly what I needed. I read over the documentation a bit fast and didn't realize you pass this into the `tablename` option. I also had some misunderstanding of how subqueries work. 

Given my example above the follow works perfectly for retrieving values out of a JSONB column.

`postgis_settings.table = '(select geom, (censusvars->>\'DP04_0009E\')::numeric as metric from table_name as foo';
`

And the filter:

`<Filter>[metric] &gt; 0</Filter>`

Some important things to note. You need to use `->>` not `->` when pulling values out of the JSONB column. The reason we need `->>` is because this returns the value as text which we can cast with `::numeric`. `->` returns JSON type which can not be cast to a numeric col.

**Comment by TehNrd** [comment]
This is a bit off topic but figured someone may find this useful if the are looking into this type of functionality and find this issue.

JSONB appears to be significantly faster in retrieving data than storing lots of data in columns.

I created two tables. One has 1200 columns. The other has 15 columns, one of which is a JSONB column. In this JSONB column there is a very simple object with key->value pairs, but it is quite large with 1200 entries, the same data as in the columns.

Query the JSONB column as described in the previous comment is approximately 3x faster than trying to query the table with 1200 columns. Apologies if this is too far off topic but seems relevant given this is a good solution/work around to this request.

**Comment by springmeyer** [comment]
Wow, awesome results @TehNrd and thanks for the details.

**Comment by springmeyer** [comment]
I'd love to see a sample github repo + hello world for how to use it with Mapnik. Maybe something like https://github.com/springmeyer/couchdb-mapnik

**Comment by DemarsM** [comment]
+1 for this feature. 

I would like to use it at the Label level. If it would be possible to use like this

    <TextSymbolizer ...>[hstore_field->'key']</TextSymbolizer>

that would be better/simpler than subquery because some features in table could have different keys in the hstore field. The subquery might be in error if the key does not exists for a feature?

**Comment by plepe** [comment]
The subquery does not throw an error if the hstore does not contain the key:
```sql
select ('foo=>bar'::hstore)->'somekey' as "somekey";
```

The column `somekey` will contain the value `null`.

**Comment by jkroll20** [comment]
I'm going to work on implementing this. I have experience with C++ but haven't worked with Mapnik yet. Currently I'm trying to wrap my head around Mapnik and related infrastructure, so I could use some help getting started. 
If someone could create a unit test/testcase for this, that would help a lot.

**Comment by brandonreavis** [comment]
Like @jkroll20, I also would love to see this implemented sometime and wouldn't mind helping, but figuring out Mapnik's inner workings is no trivial task. I think Mapnik would really benefit from detailed documentation geared for future contributors. Something explaining what each component of the source does and describing how and where it is used would help a ton. Perhaps a separate issue should be created for this.

**Comment by nyurik** [comment]
@jkroll20 and I had a long discussion about this issue today. A simple unit test would help this get started much faster, e.g. it could fake input with a complex value and check that it gets saved correctly in the vector format.

From my "black box" understanding, Mapnik needs to be updated to support "complex value" types in the internal data pipeline. Postgress plugin will detect that the resultset's field from a SQL query is an hstore or JSONB, and create a complex value. I think image creation might be a bit too complex to tackle from the start, because that would involve modifying style parser to handle <Filter>[tags=>amenity]</> and <TextSymbolizer>, but vector tile output should be much easier to implement because it is a straight "dump all fields" operation.

**Comment by jkroll20** [comment]
I am revisiting this now, and am about to work on this as a Wikimedia Germany employee. I talked to my colleagues, and we would like some additional feedback from you, that is, the people who want this feature implemented:

 * Use cases: Could you clear up what this feature would be used for, in terms of things visible to end-users
 * Could you describe what this feature would make possible, compared to existing code, i.e., why existing methods/solutions don't work.
 * Test case: Can somebody write a simple test case which would use the feature. This is essential so I can check whether my implementation actually does what's required.

We hope to get this feedback until the end of July. Thanks!

**Comment by springmeyer** [comment]
@jkroll20 - did you end up making any progress on this?

**Comment by jkroll20** [comment]
No, I stopped working on it.

**Comment by nyurik** [comment]
I have written a JavaScript workaround for this, so that the data source (Postgres) can generate all values as a single stringified JSON object -- (key-value pairs, e.g.  `language codes -> values`), and once the tile is generated by Mapnik, the library would parse the tile and substitute that single key with multiple key-values.  This way,  `{id: 1, type: road, name: {en: 'A', fr: 'B'}}` would become `{id: 1, type: road, name_en: 'A', name_fr: 'B'}`.

https://github.com/kartotherian/babel
